cfgov/cfgov/urls.py

```diff
@@ -618,18 +619,39 @@ def handle_error(code, request, exception=None):
                             "HTTP Error %s." % str(code), status=code)
 
 
-# Handle case-insensitive and extraneous characters at end of URLs
-# i.e. URLs that end in %20) have these two characters removed
-# Using (?i) in url() patterns is deprecated in Django 2.1
 def handle_404_error(code, request, exception=None):
-    extraneous_char_re = re.compile(r'[ !#$%&()*+,-.:;<=>?@\[\]^_`{|}~]+$')
+    """Attempt to self-heal 404-ing URLs.
 
-    if request.path != request.path.lower():
-        return redirect(request.path.lower(), permanent=True)
+    Takes a 404ing request and tries to transform it to a successful request by
+    lowercasing the path and stripping extraneous characters from the end.
+    If successful, it redirects to the working URL. If unsuccessful, it returns
+    the expected 404 for the original requested path.
+    """
+
+    # Lowercase the path.
+    path = request.path.lower()
+
+    # Check for and remove extraneous characters at the end of the path.
+    extraneous_char_re = re.compile(
+        r'[`~!@#$%^&*()\-_–—=+\[\]{}\\|;:\'‘’"“”,.…<>? ]+$'
+    )
+    path = re.sub(extraneous_char_re, '', path)
+
+    # If the path has changed, try resolving the new path.
+    if path != request.path:
+        try:
+            # Add trailing slash if not present, because resolve() doesn't go
```

Comment:
We have some rare URLs that don't end in a slash, say `/sitemap.xml`. If someone accesses `/sitemap.xml!`, this logic will try resolving `sitemap.xml/` and fail, instead of redirecting to `/sitemap.xml`.

Would it be better to try the resolution twice, first without a slash, and second with?

Additionally, the addition of the trailing slash is governed by [`settings.APPEND_SLASH`](https://docs.djangoproject.com/en/3.0/ref/settings/#append-slash); we do happen to have this set to true, but I wonder if it would be better to check that setting here for consistency.

Reply:
Yesss good call on the filenames at root.

I thought about checking the setting, and you're right, it's probably a good idea. Thanks for saving me from my laziness :)